File: app/src/main/java/ch/bailu/aat/views/BusyViewControl.java
package ch.bailu.aat.views;

import android.util.SparseBooleanArray;
import android.view.View;
import android.view.ViewGroup;

public class BusyViewControl {

    private static final int DEFAULT_ID=-1;

    public final BusyViewContainer busy;
    private SparseBooleanArray isWaiting = new SparseBooleanArray(5);



    public BusyViewControl (ViewGroup parent) {
        busy = new BusyViewContainer(parent.getContext());
        parent.addView(busy,
                new ViewGroup.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

        busy.setVisibility(View.GONE);
    }

    public void startWaiting() {
        startWaiting(DEFAULT_ID);
    }

    public void stopWaiting() {
        stopWaiting(DEFAULT_ID);
    }

    public void startWaiting(int id) {
        changeWaiting(id, true);
    }

    public void stopWaiting(int id) {
        changeWaiting(id, false);
    }


    private void changeWaiting(int id, boolean w) {
        isWaiting.put(id, w);

        if (isWaiting()) {
            busy.setVisibility(View.VISIBLE);
            busy.bringToFront();
        } else {
            busy.setVisibility(View.GONE);
        }

    }

    public boolean isWaiting() {
        for (int i=0; i<isWaiting.size(); i++) {
            if (isWaiting.valueAt(i)) {
                return true;
            }
        }
        return false;
    }



    public void setText(CharSequence t) {
        busy.setText(t);
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/views/busy/BusyViewControl.kt
package ch.bailu.aat.views.busy

import android.view.View
import android.view.ViewGroup
import ch.bailu.aat_lib.util.IndexedMap

open class BusyViewControl(parent: ViewGroup) {
    val busy: BusyViewContainer
    private val busyMap = IndexedMap<Int, Boolean>()

    init {
        busy = BusyViewContainer(parent.context)
        parent.addView(
            busy,
            ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
            )
        )
        busy.visibility = View.GONE
    }

    @JvmOverloads
    fun startWaiting(id: Int = DEFAULT_ID) {
        changeWaiting(id, true)
    }

    @JvmOverloads
    fun stopWaiting(id: Int = DEFAULT_ID) {
        changeWaiting(id, false)
    }

    private fun changeWaiting(id: Int, w: Boolean) {
        busyMap.put(id, w)
        if (isWaiting()) {
            busy.visibility = View.VISIBLE
            busy.bringToFront()
        } else {
            busy.visibility = View.GONE
        }
    }

    fun isWaiting(): Boolean {
        for (i in 0 until busyMap.size()) {
            if (true == busyMap.getValueAt(i)) {
                return true
            }
        }
        return false
    }

    fun setText(text: CharSequence) {
        busy.setText(text)
    }

    companion object {
        private const val DEFAULT_ID = -1
    }
}
